File: sdk/clojure/src/rocksky/spotify.clj
(ns rocksky.spotify
  "Endpoints under app.rocksky.spotify.* — Spotify-backed remote control."
  (:require [rocksky.client :as c]))

(defn get-currently-playing
  "Get the currently playing track. Optional: `:actor`."
  ([client] (get-currently-playing client nil))
  ([client {:keys [actor]}]
   (c/query client :app.rocksky.spotify.getCurrentlyPlaying {:actor actor})))

(defn play
  "Resume playback."
  [client]
  (c/procedure client :app.rocksky.spotify.play))

(defn pause
  "Pause playback."
  [client]
  (c/procedure client :app.rocksky.spotify.pause))

(defn next-track
  "Skip to the next track."
  [client]
  (c/procedure client :app.rocksky.spotify.next))

(defn previous-track
  "Skip to the previous track."
  [client]
  (c/procedure client :app.rocksky.spotify.previous))

(defn seek
  "Seek to `:position` (in seconds). Required: `:position`."
  [client {:keys [position]}]
  (c/procedure client :app.rocksky.spotify.seek nil {:position position}))

File: sdk/elixir/lib/rocksky/spotify.ex
defmodule Rocksky.Spotify do
  @moduledoc "`app.rocksky.spotify.*` endpoints — Spotify remote control for the authenticated user."

  alias Rocksky.HTTP

  @doc "Currently playing on Spotify. Params: `:actor`."
  def get_currently_playing(client, params \\ []),
    do: HTTP.query(client, "app.rocksky.spotify.getCurrentlyPlaying", params)

  @doc "Resume Spotify playback."
  def play(client, params \\ []),
    do: HTTP.procedure(client, "app.rocksky.spotify.play", params)

  @doc "Pause Spotify playback."
  def pause(client, params \\ []),
    do: HTTP.procedure(client, "app.rocksky.spotify.pause", params)

  @doc "Skip to next track."
  def next(client, params \\ []),
    do: HTTP.procedure(client, "app.rocksky.spotify.next", params)

  @doc "Go to previous track."
  def previous(client, params \\ []),
    do: HTTP.procedure(client, "app.rocksky.spotify.previous", params)

  @doc "Seek to position (ms). Params: `:position`."
  def seek(client, params \\ []),
    do: HTTP.procedure(client, "app.rocksky.spotify.seek", params)
end
